metricbeat/module/vsphere/virtualmachine/virtualmachine.go

```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+		MaxSample:  1,
+	}
 
-		totalMemory := int64(vm.Summary.Config.MemorySizeMB) * 1024 * 1024
-		if totalMemory > 0 {
-			freeMemory := totalMemory - usedMemory
-			// Avoid negative values if reported used memory is slightly over total configured.
-			if freeMemory < 0 {
-				freeMemory = 0
-			}
-			event.Put("memory.total.guest.bytes", totalMemory)
-			event.Put("memory.free.guest.bytes", freeMemory)
-		}
+	// Create a slice to hold our query specifications for each VM
+	var metricsToQuery []string
+
+	// Iterate over the counterInfo slice
+	for _, counter := range counterInfo {
+		// Check if the counter is collected at a 20-second interval
+		if counter.Level == int32(m.MetricLevel) {
+			// Add the counter to the metricsToQuery slice
+			metricsToQuery = append(metricsToQuery, counter.Name())
 
-		if host := vm.Summary.Runtime.Host; host != nil {
-			event["host.id"] = host.Value
-			hostSystem, err := getHostSystem(ctx, c, host.Reference())
-			if err == nil {
-				event["host.hostname"] = hostSystem.Summary.Config.Name
-			} else {
-				m.Logger().Debug(err.Error())
-			}
-		} else {
-			m.Logger().Debug("'Host', 'Runtime' or 'Summary' data not found. This is either a parsing error " +
-				"from vsphere library, an error trying to reach host/guest or incomplete information returned " +
-				"from host/guest")
 		}
+	}
 
-		// Get custom fields (attributes) values if get_custom_fields is true.
-		if m.GetCustomFields && vm.Summary.CustomValue != nil {
-			customFields := getCustomFields(vm.Summary.CustomValue, customFieldsMap)
+	vmList, err := m.getVirtualMachineList(ctx, c, mgr, rpList)
+	if err != nil {
+		return fmt.Errorf("error getting Virtual Machine List: %w", err)
+	}
 
-			if len(customFields) > 0 {
-				event["custom_fields"] = customFields
-			}
-		} else {
-			m.Logger().Debug("custom fields not activated or custom values not found/parse in Summary data. This " +
-				"is either a parsing error from vsphere library, an error trying to reach host/guest or incomplete " +
-				"information returned from host/guest")
+	vmChunks := m.getVmChunks(vmList)
+	vmMap := getVmMap(vmList)
+
+	for _, chunk := range vmChunks {
+		// Perform the query for all VMs at once
+		metricBase, err := perfManager.SampleByName(ctx, querySpec, metricsToQuery, chunk)
+		if err != nil {
+			fmt.Printf("Error retrieving metrics: %s\n", err)
+			return nil // this probably needs to be a different error
 		}
 
-		if vm.Summary.Vm != nil {
-			networkNames, err := getNetworkNames(ctx, c, vm.Summary.Vm.Reference())
-			if err != nil {
-				m.Logger().Debug(err.Error())
-			} else {
-				if len(networkNames) > 0 {
-					event["network_names"] = networkNames
+		// assign metrics to the event output
+		for _, base := range metricBase {
+			if metric, ok := base.(*types.PerfEntityMetric); ok {
```

Comment:
anti pattern: see this example: https://www.jetbrains.com/help/resharper/InvertIf.html

Invert the condition and code would look cleaner.

Reply:
appreciate the feedback on this one, got the code updated to invert the condition